dbms/src/Interpreters/Join.cpp

```diff
@@ -1621,15 +1677,40 @@ void Join::joinBlockImpl(Block & block, const Maps & maps) const
         block.insert(ColumnWithTypeAndName(std::move(added_columns[i]), sample_col.type, sample_col.name));
     }
 
-    /// If ANY INNER | RIGHT JOIN - filter all the columns except the new ones.
-    if (filter && !(kind == ASTTableJoin::Kind::Anti && strictness == ASTTableJoin::Strictness::All))
-        for (size_t i = 0; i < existing_columns; ++i)
-            block.safeGetByPosition(i).column = block.safeGetByPosition(i).column->filter(*filter, -1);
+    size_t process_rows = probe_process_info.end_row - probe_process_info.start_row;
 
-    /// If ALL ... JOIN - we replicate all the columns except the new ones.
-    if (offsets_to_replicate)
-        for (size_t i = 0; i < existing_columns; ++i)
-            block.safeGetByPosition(i).column = block.safeGetByPosition(i).column->replicate(*offsets_to_replicate);
+    // if rows equal 0 means that it call getHeader, we could ignore filter and offsets_to_replicate, and do not need to update start row.
```

Comment:
nit: Because we may join an empty block, though it is not `getHeader`
```suggestion
    // if rows equal 0, we could ignore filter and offsets_to_replicate, and do not need to update start row.
```

Reply:
> nit: Because we may join an empty block, though it is not `getHeader`

done